HACKING.md

```diff
@@ -4,70 +4,84 @@ Hacking on `snapd` is fun and straightforward. The code is extensively unit
+>If you want to build or test on older versions of Ubuntu, additional steps
 may be required when installing dependencies.
 
 
 ### Supported Go version
 
-Go 1.13 or later is required to build `snapd`.
+[Go 1.13](https://pkg.go.dev/go@go1.13) or later is required to build `snapd`.
 
-If you need to build older versions of snapd, please have a look at the file
-`debian/control` to find out what dependencies were needed at the time
+>If you need to build older versions of snapd, please have a look at the file
+[debian/control](debian/control) to find out what dependencies were needed at the time
 (including which version of the Go compiler).
 
-### Setting up your build environment
+### Setting up build environment
 
-If your Go environment (e.g. `GOPATH`) is already configured, you should skip
-this step. The Go environment setup can be added to your shell login script
-(e.g. `~/.bashrc` for bash) for automatic setup (after a terminal restart).
+`snapd` is published and maintained as a [Go module](https://go.dev/doc/modules/managing-source). 
+A module is a collection of [Go packages](https://go.dev/ref/spec#Packages) 
+stored in a file tree with a [go.mod](go.mod) file at its root. 
+The [go.mod](go.mod) file defines the module’s _module path_, 
+which is also the import path used for the root directory, 
+and its _dependency requirements_. 
+All this allows maintaining `snapd` outside of the strict hierarchy of `GOPATH`, 
+that was mandatory before the release of [Go 1.13](https://pkg.go.dev/go@go1.13)
+(see [build @Go1.10](https://pkg.go.dev/go/build@go1.10)).  
+
+Even with Go modules mode, `GOPATH` is still used by the Go tool implicitly, 
+to store downloaded source code and compiled commands 
+(see [Go docs](https://pkg.go.dev/cmd/go#hdr-GOPATH_and_Modules)).
+
+> If `$GOPATH` is not set explicitly, in Unix-like systems it defaults to `$HOME/go`.
+
+One can easily find the location of the `GOPATH` file-tree on any enabled machine
+by running the following command:
 
 ```
-export GOPATH=${HOME}/work
-mkdir -p $GOPATH
-export PATH="$PATH:$GOPATH/bin"
+go env GOPATH
 ```
 
-### Further environment variable details
-
-When working with the source of Go programs, you should define a path within
-your home directory (or other workspace) which will be your `GOPATH`. `GOPATH`
-is similar to Java's `CLASSPATH` or Python's `~/.local`. `GOPATH` is
-documented [online](http://golang.org/pkg/go/build/) and inside the `go` tool
-itself.
+#### Backward compatibility
 
-    go help gopath
+Some tools we use still rely on `$GOPATH/bin` being listed in `$PATH` 
+(see [Running spread with QEMU](#running-spread-with-qemu)).
+Add `$GOPATH/bin` to your `PATH` in your shell login script 
+(e.g. `~/.bashrc` for bash):
 
-Various conventions exist for naming the location of your `GOPATH`, but it
-should exist, and be writable by you. For example:
+```sh
+export PATH=$PATH:$(go env GOPATH)/bin
```

Comment:
Mmm... but where does `go env` get the value from? If it still uses `GOPATH` then using the environment variable directly will be much faster (we should try to keep the user's `.bashrc` execution as quick as possible).

Reply:
This is exactly the change that comes with Go1.13. `go` tool doesn't rely on `GOPATH` variable. It has a default value for this parameter, which can be overwritten with `GOPATH` environment variable. However, this method (i.e. environment variable), per my understanding of the Go documentation, is considered deprecated and is maintained primarily to support older projects that started before Go1.13 and haven't (or couldn't) migrate to the new mode of work (i.e. Modules).
By the way, it has no effect on performance whether you have this environment variable defined or not. Moreover, as I understood this variable might have negative effect (if explicitly defined) on certain modern tools that anticipate things in the default location (bad excuse - I agree, but this stuff happens).

Reply:
Sorry, I was not very clear about the performance issue: I mean that the `go env` command takes time to execute, and if you put this in the `.bashrc` it will slow down the prompt. By a few milliseconds only, sure, but some time ago I had to go through my `.bashrc` and remove a few entries because it had grown to the point where the delay was not just noticeable, but annoying.

Anyway, this is not a big issue.